src/m3ninx/index/regexp.go

```diff
@@ -0,0 +1,323 @@
+	ast := &syntax.Regexp{
+		Op:    syntax.OpConcat,
+		Flags: syntax.Perl,
+		Sub: []*syntax.Regexp{
+			&syntax.Regexp{
+				Op:    syntax.OpBeginText,
+				Flags: syntax.Perl,
+			},
+			unanchoredRegexp,
+			&syntax.Regexp{
+				Op:    syntax.OpEndText,
+				Flags: syntax.Perl,
+			},
+		},
+	}
+	return simplify(ast.Simplify()), nil
+}
+
+// ensureRegexpUnanchored strips '^' and '$' characters from appropriate locations in the parsed syntax.Regexp,
+// to ensure every input regular expression is converted to it's equivalent un-anchored regular expression
+// assuming the entire input is matched.
+func ensureRegexpUnanchored(parsed *syntax.Regexp) (*syntax.Regexp, error) {
+	r, _, err := ensureRegexpUnanchoredHelper(parsed, true, true)
+	if err != nil {
+		return nil, err
+	}
+	return simplify(r), nil
+}
+
+func ensureRegexpUnanchoredHelper(parsed *syntax.Regexp, leftmost, rightmost bool) (output *syntax.Regexp, changed bool, err error) {
+	// short circuit when we know we won't make any changes to the underlying regexp.
+	if !leftmost && !rightmost {
+		return parsed, false, nil
+	}
+
+	newRegexp, changed := parsed, false
+	switch parsed.Op {
+	case syntax.OpBeginLine, syntax.OpEndLine:
+		// i.e. the flags provided to syntax.Parse did not include the `OneLine` flag, which
+		// should never happen as we're using syntax.Perl which does include it (ensured by a test
+		// in this package).
+		return nil, false, fmt.Errorf("regular expressions are forced to be single line")
+	case syntax.OpBeginText:
+		if leftmost {
+			newRegexp = &syntax.Regexp{
+				Op:    syntax.OpEmptyMatch,
+				Flags: parsed.Flags,
+			}
+			changed = true
+		}
+	case syntax.OpEndText:
+		if rightmost {
+			newRegexp = &syntax.Regexp{
+				Op:    syntax.OpEmptyMatch,
+				Flags: parsed.Flags,
+			}
+			changed = true
+		}
+	case syntax.OpCapture:
+		// because golang regexp's don't allow backreferences, we don't care about maintaining capture
+		// group namings and can treate captures the same as we do conactenations.
+		fallthrough
+	case syntax.OpConcat:
+		// strip left-most '^'
+		if l := len(parsed.Sub); leftmost && l > 0 {
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost && l == 1)
+			if err != nil {
+				return nil, false, err
+			}
+			if c {
+				parsed.Sub[0] = newRe
+				changed = true
+			}
+		}
+		// strip right-most '$'
+		if l := len(parsed.Sub); rightmost && l > 0 {
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[l-1], leftmost && l == 1, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if c {
+				parsed.Sub[l-1] = newRe
+				changed = true
+			}
+		}
+	case syntax.OpAlternate:
+		// strip left-most '^' and right-most '$' in each sub-expression
+		for idx := range parsed.Sub {
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[idx], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			parsed.Sub[idx] = newRe
+			changed = changed || c
```

Comment:
super nit: Since the other cases do `if c { parsed.Sub[...] = ...; changed = true }` perhaps makes sense to follow that pattern here too?

Reply:
sure